ClosedXML/Excel/Ranges/XLRangeBase.cs

```diff
@@ -1557,39 +1561,64 @@ public void Delete(XLShiftDeletedCells shiftDeleteCells)
+                    cellsQuery = Worksheet.Internals.CellsCollection.GetCells(
+                        RangeAddress.FirstAddress.RowNumber,
+                        RangeAddress.FirstAddress.ColumnNumber,
+                        RangeAddress.LastAddress.RowNumber,
+                        Worksheet.Internals.CellsCollection.MaxColumnUsed);
+
+                    columnModifier = ColumnCount();
+
+                    break;
+
+                case XLShiftDeletedCells.ShiftCellsUp:
+                    cellsQuery = Worksheet.Internals.CellsCollection.GetCells(
+                        RangeAddress.FirstAddress.RowNumber,
+                        RangeAddress.FirstAddress.ColumnNumber,
+                        Worksheet.Internals.CellsCollection.MaxRowUsed,
+                        RangeAddress.LastAddress.ColumnNumber);
+
+                    rowModifier = RowCount();
 
-            int columnModifier = shiftDeleteCells == XLShiftDeletedCells.ShiftCellsLeft ? ColumnCount() : 0;
-            int rowModifier = shiftDeleteCells == XLShiftDeletedCells.ShiftCellsUp ? RowCount() : 0;
-            var cellsQuery = shiftDeleteCells == XLShiftDeletedCells.ShiftCellsLeft ? shiftLeftQuery : shiftUpQuery;
-            foreach (XLCell c in cellsQuery)
+                    break;
+
+                default:
+                    cellsQuery = new XLCell[] { };
+                    break;
+            }
+
+            foreach (var c in cellsQuery)
             {
-                var newKey = new XLAddress(Worksheet, c.Address.RowNumber - rowModifier,
-                                           c.Address.ColumnNumber - columnModifier,
-                                           false, false);
-                var newCell = new XLCell(Worksheet, newKey, c.StyleValue);
-                newCell.CopyValuesFrom(c);
-                newCell.FormulaA1 = c.FormulaA1;
+                // Schedule for removal from CellsCollection
                 cellsToDelete.Add(c.Address);
 
-                bool canInsert = shiftDeleteCells == XLShiftDeletedCells.ShiftCellsLeft
-                                     ? c.Address.ColumnNumber > RangeAddress.LastAddress.ColumnNumber
-                                     : c.Address.RowNumber > RangeAddress.LastAddress.RowNumber;
+                // Generate new cell to insert into CellsCollection
+                var newCellAddress = new XLAddress(Worksheet, c.Address.RowNumber - rowModifier,
+                                           c.Address.ColumnNumber - columnModifier,
+                                           fixedRow: false,
+                                           fixedColumn: false);
 
-                if (canInsert)
-                    cellsToInsert.Add(newKey, newCell);
+                if (newCellAddress.IsValid)
+                {
+                    var newCell = new XLCell(Worksheet, newCellAddress, c.StyleValue);
+                    newCell.CopyValuesFrom(c);
```

Comment:
Never thought about how cells are removed during ranges shifting in ClosedXML.

Say, if the user has a reference to the cell `B2` and then removes the column `A`, he'll still have a reference to the cell which a) is not presented on the worksheet, b) has an address `B2`, c) has a value same as the new `A2` cell. And even though this cell is removed from a worksheet I am almost sure that manipulating with it will affect the actual worksheet in an unpredictable manner. This seems wrong to me. Besides, this causes an unnecessary allocation of `XLCells`, so I see the possibility for improvement here.

For now, let's leave it as it is, your refactoring made the code easier to read indeed. We can make it better some time later.

Reply:
Not sure I follow you - maybe you can illustrate with an example. But what happens currently, references to cells doesn't necessarily cause a new entry in `XLWorksheet.Internals.CellsCollection`. Only if the reference is evaluated in some way, e.g. a value is requested, or if the target cell has a style applied (or other conditions) does an entry get added. 

When column `A` is removed, cell `B2` will be removed from `CellsCollection`. A new entry with address `A2` will be created and the old `B2`'s properties will be copied to `A2`. See the `newCell.CopyValuesFrom(c);` line. So cells here are never shifted. They are explicitly removed and readded with a new address. I do see, however, how this is not optimal in terms of performance, though.

Reply:
Created #1162 to address this.